### PR TITLE
apis: make the APIs singleton classes

### PR DESCRIPTION
**Pull request by freesteph**
There's no need to hold state in APIs. They were initialised with a UAI parameter because historically they were only used to fetch an establishment's students but that's not relevant anymore, especially with endpoints that aren't related to an establishment (i.e: `/eleves/{ine}/`).

Change the APIs to be singleton classes and use a new interface that forces symetry on both APIs (FREGATA and SYGNE):

```rb
fetch_resource(:resource_type, params)
```

with :resource_type being either:

- establishment_students ;
- student ;
- student_schoolings.

This makes it easier to identify what data we try and get from both APIs (I say "try" because they have different endpoints/data topology), and for a nicer interface:

```rb
api = StudentsApi::Sygne # that's it.

api.fetch_resource :establishment_students, { uai: "007" }

api.get "/some/custom/path" # if you want to inspect another endpoint
```

which also manages to extract the ungodly mess that was `fetch_and_parse!` which was doing a "grab establishment students + map + upsert them". Now the `FetchStudentsJob` clearly separate and exhibits that mess, which is a great step towards a nice service-oriented refactor next to the new StudentSchoolingsUpdater.

### CODE REVIEW

File: app/apis/students_api/base.rb [hunk 1]
```diff
@@ -2,53 +2,53 @@
...
+      # NOTE: these are "actual" mappers, they turn specific API data
+      # projections into known hash structures.
+      %w[schooling address classe student].each do |klass|
+        define_method "#{klass}_mapper" do
+          mapper = "StudentsApi::#{identifier}::Mappers::#{klass.classify}Mapper".constantize
 
-    def fetch_and_parse!
-      mapper.new(response, uai).parse!
-    end
+          mapper.new
```
Comment (pskl) [inline app/apis/students_api/base.rb:39]: C'est un peu bizarre de faire une instantiation au moment de definir la méthode non? 

La plupart des clients API sont comme ca: 
```ruby
client = Module::APIClient.new
```
Non?

File: docs/apis_élèves.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
+# APIs pour la récupération des élèves
```
Comment (pskl) [inline docs/apis_élèves.md:1]: Excellent ca! 🍾 Merci infiniment

File: docs/apis_élèves.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+simple INE (à la connaissance de l'auteur). Comme la terminaison des
+listes d'élèves fournit l'intégralité des données nécessaires pour
+APLyPro, on utilise cette dernière pour récupérer tous les élèves puis
+aller chercher l'élève demandé.
+
+## Sécurisation des APIS
+
+SYGNE utilise OAuth2.0 pour authentifier les requêtes.
+
+FREGATA utilise HMAC.
```
Comment (pskl) [inline docs/apis_élèves.md:66]: HMAC SHA-1 en plus qui est vulnérable à des attaques nan et deprecated depuis à peu près 5 ans? Il me semble... (même si on peut pas y faire grand chose, on peut faire remonter l'info à Fregata peut-être)

File: spec/apis/students_api/fregata/api_spec.rb [hunk 4]
```diff
@@ -3,43 +3,52 @@
 require "rails_helper"
 
 describe StudentsApi::Fregata::Api do
-  subject(:api) { described_class.new(establishment.uai) }
```
Comment (pskl) [inline spec/apis/students_api/fregata/api_spec.rb:6]: Pour moi le new devrait être externe toujours, au niveau DX c'est plus intuitif (moins magique)

**Review by pskl — COMMENTED** [review]
Nice 💘

File: spec/apis/students_api/fregata/api_spec.rb (hunk 4 above)
Comment (freesteph) [reply spec/apis/students_api/fregata/api_spec.rb:6]: j'ai pas compris ? le sujet maintenant c'est la classe elle-même ?

File: app/apis/students_api/base.rb (hunk 1 above)
Comment (freesteph) [reply app/apis/students_api/base.rb]: j'ai désinstancié les mappers du coup

**Review by pskl — APPROVED** [review]

**Pull request merged**